cheddar/cheddar-application/src/test/java/com/clicktravel/cheddar/application/retry/RetryableAspectTest.java

```diff
@@ -0,0 +1,250 @@
+ *
+ */
+package com.clicktravel.cheddar.application.retry;
+
+import static com.clicktravel.common.random.Randoms.randomIntInRange;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.*;
+
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.joda.time.DateTime;
+import org.junit.Test;
+
+@SuppressWarnings("unchecked")
+public class RetryableAspectTest {
+
+    @Test
+    public void shouldConstructRetryableAspect() {
+        // Given
+        RetryableAspect retryableAspect = null;
+
+        // When
+        retryableAspect = new RetryableAspect();
+
+        // Then
+        assertNotNull(retryableAspect);
+    }
+
+    @Test
+    public void shouldAttemptMethodAndNotRetry_withExceptionThrownAndRetryDisabled() throws Throwable {
+        // Given
+        final RetryableAspect retryableAspect = new RetryableAspect();
+        final ProceedingJoinPoint mockProceedingJoinPoint = setupSimpleProceedingJoinPointMock();
+        final Retryable mockRetryable = setupSimpleRetryableMock();
+
+        RetryableConfiguration.setRetryableEnabled(false);
+        when(mockProceedingJoinPoint.proceed()).thenThrow(new RetryAspectTestException());
+
+        // When
+        RetryAspectTestException actualException = null;
+        try {
+            retryableAspect.attemptMethodAndRetryIfNeeded(mockProceedingJoinPoint, mockRetryable);
+        } catch (final RetryAspectTestException e) {
+            actualException = e;
+        }
+
+        // Then
+        verify(mockProceedingJoinPoint).proceed();
+
+        assertNotNull(actualException);
+    }
+
+    @Test
+    public void shouldAttemptMethodAndNotRetry_withNoExceptionThrownAndRetryEnabled() throws Throwable {
+        // Given
+        final RetryableAspect retryableAspect = new RetryableAspect();
+        final ProceedingJoinPoint mockProceedingJoinPoint = setupSimpleProceedingJoinPointMock();
+        final Retryable mockRetryable = setupSimpleRetryableMock();
+
+        RetryableConfiguration.setRetryableEnabled(true);
+
+        // When
+        Exception actualException = null;
+        try {
+            retryableAspect.attemptMethodAndRetryIfNeeded(mockProceedingJoinPoint, mockRetryable);
+        } catch (final RetryAspectTestException e) {
+            actualException = e;
+        }
+
+        // Then
+        verify(mockProceedingJoinPoint).proceed();
+
+        assertNull(actualException);
+    }
+
+    @Test
+    public void shouldAttemptMethodAndRetry_withExceptionThrownAndRetryEnabledAndMaxAttemptsAdjusted()
+            throws Throwable {
+        // Given
+        final RetryableAspect retryableAspect = new RetryableAspect();
+        final ProceedingJoinPoint mockProceedingJoinPoint = setupSimpleProceedingJoinPointMock();
+        final int retryMaxAttempts = randomIntInRange(2, 4);
```

Comment:
This variable is wrongly named, it should be `maxAttempts`. This value defines the number of times the advised method is *attempted*.
`maxAttempts = 1` means the method will not be retried.
`maxAttempts = 2` means the method will be retried at most once.
`maxAttempts = 3` means the method will be retried at most twice, etc